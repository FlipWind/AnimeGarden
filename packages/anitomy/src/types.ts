import type { ElementCategory } from './element';

export interface AnitomyOptions {
  delimiters: string;

  parseEpisodeNumber: boolean;

  parseEpisodeTitle: boolean;

  parseFileExtension: boolean;

  parseReleaseGroup: boolean;
}

export type ParsedResult = Partial<Record<ElementCategory, string>>;

export interface AnitomyResult {
  title: string | undefined;

  type: string | undefined;

  season: string | undefined;

  year: string | undefined;

  month: string | undefined;

  language: string | undefined;

  subtitles: string | undefined;

  source: string | undefined;

  episode: {
    number: string | undefined;

    numberAlt: string | undefined;

    title: string | undefined;
  };

  volume: {
    number: string | undefined;
  };

  video: {
    term: string | undefined;

    resolution: string | undefined;
  };

  audio: {
    term: string | undefined;
  };

  file: {
    name: string;

    checksum: string | undefined;

    extension: string | undefined;
  };

  // prefix: {
  //   season: string | undefined;
  //   volume: string | undefined;
  //   episode: string | undefined;
  // };
}
