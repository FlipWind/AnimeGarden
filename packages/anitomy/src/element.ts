export type ElementCategory =
  | 'season'
  | 'seasonPrefix'
  | 'title'
  | 'type'
  | 'year'
  | 'episodeNumber'
  | 'episodePrefix'
  | 'episodeTitle'
  | 'language'
  | 'releaseGroup'
  | 'releaseInformation'
  | 'releaseVersion'
  | 'source'
  | 'subtitles'
  // Audio and Video
  | 'audioTerm'
  | 'videoResolution'
  | 'videoTerm'
  // Volume
  | 'volumeNumber'
  | 'volumePrefix'
  // File related
  | 'checksum'
  | 'extension'
  | 'filename'
  // Other
  | 'other'
  | 'unknown';
