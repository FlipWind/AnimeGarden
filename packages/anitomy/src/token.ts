export enum TokenCategory {
  Unknown = 'Unknown',
  Bracket = 'Bracket',
  Delimiter = 'Delimiter',
  Identifier = 'Identifier',
  Invalid = 'Invalid'
}

export interface Token {
  category: TokenCategory;

  content: string;

  enclosed: boolean;
}

export class TextRange {
  public text: string;

  public offset: number;

  public size: number;

  public constructor(text: string, offset: number, size: number) {
    this.text = text;
    this.offset = offset;
    this.size = size;
  }

  public fork(offset: number, size: number) {
    return new TextRange(this.text, offset, size);
  }

  public toString() {
    return this.text.slice(this.offset, this.offset + this.size);
  }
}
