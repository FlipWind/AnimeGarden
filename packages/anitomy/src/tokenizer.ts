import type { AnitomyOptions, ParsedResult } from './types';

export function tokenize(filename: string, result: ParsedResult, options: AnitomyOptions) {
  const tokens: Token[] = [];
  return tokens;
}

export class Token {}
