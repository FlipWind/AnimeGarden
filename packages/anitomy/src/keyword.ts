import { ElementCategory } from './element';

export interface KeywordOptions {
  identifiable: boolean;

  searchable: boolean;

  valid: boolean;
}

export type Keyword = {
  category: ElementCategory;
} & KeywordOptions;

export class KeywordManager {
  private static keys = new Map<string, Keyword>();

  private static extensions = new Map<string, Keyword>();

  static {
    const optionsDefault: KeywordOptions = { identifiable: true, searchable: true, valid: true };
    const optionsInvalid: KeywordOptions = { identifiable: false, searchable: false, valid: false };
    const optionsUnidentifiable: KeywordOptions = {
      identifiable: false,
      searchable: true,
      valid: true
    };
    const optionsUnidentifiableInvalid: KeywordOptions = {
      identifiable: false,
      searchable: true,
      valid: false
    };
    const optionsUnidentifiableUnsearchable: KeywordOptions = {
      identifiable: false,
      searchable: false,
      valid: true
    };

    add('seasonPrefix', optionsUnidentifiable, ['SAISON', 'SEASON']);
    add('type', optionsUnidentifiable, [
      'GEKIJOUBAN',
      'MOVIE',
      'OAD',
      'OAV',
      'ONA',
      'OVA',
      'SPECIAL',
      'SPECIALS',
      'TV'
    ]);
    add('type', optionsUnidentifiableUnsearchable, ['SP']);
    add('type', optionsUnidentifiableInvalid, [
      'ED',
      'ENDING',
      'NCED',
      'NCOP',
      'OP',
      'OPENING',
      'PREVIEW',
      'PV'
    ]);
    add('audioTerm', optionsDefault, [
      // Audio channels
      '2.0CH',
      '2CH',
      '5.1',
      '5.1CH',
      'DTS',
      'DTS-ES',
      'DTS5.1',
      'TRUEHD5.1',
      // Audio codec
      'AAC',
      'AACX2',
      'AACX3',
      'AACX4',
      'AC3',
      'EAC3',
      'E-AC-3',
      'FLAC',
      'FLACX2',
      'FLACX3',
      'FLACX4',
      'LOSSLESS',
      'MP3',
      'OGG',
      'VORBIS',
      // Audio language
      'DUALAUDIO',
      'DUAL AUDIO'
    ]);
    add('videoTerm', optionsDefault, [
      // Frame rate
      '23.976FPS',
      '24FPS',
      '29.97FPS',
      '30FPS',
      '60FPS',
      '120FPS',
      // Video codec
      '8BIT',
      '8-BIT',
      '10BIT',
      '10BITS',
      '10-BIT',
      '10-BITS',
      'HI10',
      'HI10P',
      'HI444',
      'HI444P',
      'HI444PP',
      'H264',
      'H265',
      'H.264',
      'H.265',
      'X264',
      'X265',
      'X.264',
      'AVC',
      'HEVC',
      'HEVC2',
      'DIVX',
      'DIVX5',
      'DIVX6',
      'XVID',
      // Video format
      'AVI',
      'RMVB',
      'WMV',
      'WMV3',
      'WMV9',
      // Video quality
      'HQ',
      'LQ',
      // Video resolution
      'HD',
      'SD'
    ]);

    add('episodePrefix', optionsDefault, [
      'EP',
      'EP.',
      'EPS',
      'EPS.',
      'EPISODE',
      'EPISODE.',
      'EPISODES',
      'CAPITULO',
      'EPISODIO',
      'FOLGE'
    ]);
    add('episodePrefix', optionsInvalid, ['E', '\\x7B2C']);

    add('extension', optionsDefault, [
      '3GP',
      'AVI',
      'DIVX',
      'FLV',
      'M2TS',
      'MKV',
      'MOV',
      'MP4',
      'MPG',
      'OGM',
      'RM',
      'RMVB',
      'TS',
      'WEBM',
      'WMV'
    ]);
    add('extension', optionsInvalid, [
      'AAC',
      'AIFF',
      'FLAC',
      'M4A',
      'MP3',
      'MKA',
      'OGG',
      'WAV',
      'WMA',
      '7Z',
      'RAR',
      'ZIP',
      'ASS',
      'SRT'
    ]);

    function add(category: ElementCategory, options: KeywordOptions, keys: string[]) {
      const map = KeywordManager.container(category);
      for (const key of keys) {
        if (!map.has(key)) {
          map.set(key, { category, ...options });
        }
      }
    }
  }

  public static contains(category: ElementCategory, keyword: string) {
    const map = KeywordManager.container(category);
    const value = map.get(keyword);
    return value && value.category === category;
  }

  public static container(category: ElementCategory) {
    return category === 'extension' ? KeywordManager.extensions : KeywordManager.keys;
  }
}
