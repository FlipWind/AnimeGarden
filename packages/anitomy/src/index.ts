import type { AnitomyOptions, ParsedResult } from './types';

import { tokenize } from './tokenizer';
import { mergeResult } from './utils';
import { KeywordManager } from './keyword';

export type * from './types';

export function parse(filename: string, _options: Partial<AnitomyOptions> = {}): ParsedResult {
  if (filename === '') return {};

  let result: ParsedResult = {};

  const options = resolveOptions(_options);

  result.filename = filename;
  if (options.extension) {
    const ext = removeExtension(filename);
    if (ext) {
      result.filename = filename;
      result.extension = ext.extension;
    }
  }

  const tokenized = tokenize(result.filename!, options);
  result = mergeResult(result, tokenized.result);
  if (!tokenized.ok) {
    return result;
  }

  return result;
}

export function resolveOptions(options: Partial<AnitomyOptions>): AnitomyOptions {
  return {
    delimiters: ' _.&+,|',
    episode: true,
    extension: true,
    ...options
  };
}

function removeExtension(filename: string) {
  const position = filename.lastIndexOf('.');
  if (position === -1) return undefined;

  const extension = filename.slice(position + 1);
  if (extension.length > 4 || !/^[a-zA-Z0-0]+$/.test(extension)) {
    return undefined;
  }

  if (!KeywordManager.contains('extension', extension)) {
    return undefined;
  }

  return {
    filename: filename.slice(0, position),
    extension
  };
}
