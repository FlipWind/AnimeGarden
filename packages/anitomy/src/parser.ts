import type { AnitomyOptions, ParsedResult } from './types';

import { isNumericString, trim } from './utils';
import { Token, TokenCategory } from './token';
import { KeywordManager } from './keyword';

export function parse(tokens: Token[], options: AnitomyOptions) {
  const result: ParsedResult = {};

  searchForKeywords();
  searchForIsolatedNumbers();

  if (options.episode) {
    searchForEpisodeNumber();
  }

  searchForAnimeTitle();

  return { result };

  function searchForKeywords() {
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      if (token.category !== TokenCategory.Unknown) continue;

      const word = trim(token.content, [' ', '-']);
      if (word === '') continue;

      // Don't bother if the word is a number that cannot be CRC
      if (word.length !== 8 && isNumericString(word)) continue;

      const keyword = KeywordManager.normalize(word);
      
    }
  }

  function searchForIsolatedNumbers() {}

  function searchForEpisodeNumber() {}

  function searchForAnimeTitle() {}
}
