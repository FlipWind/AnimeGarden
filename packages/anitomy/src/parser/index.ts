import type { AnitomyOptions, ParsedResult } from '../types';

import { KeywordManager } from '../keyword';
import { ElementCategory } from '../element';
import { Token, TokenCategory } from '../token';
import { isNumericString, trim } from '../utils';

import {
  isCRC32,
  isElementCategorySearchable,
  isElementCategorySingular,
  isResolution
} from './utils';
import { AnimeYearMax, AnimeYearMin } from './number';
import { ParserContext, hasResult, setResult } from './context';
import { CheckAndSetAnimeSeasonKeyword, checkExtentKeyword, isTokenIsolated } from './parser';

export function parse(result: ParsedResult, tokens: Token[], options: AnitomyOptions) {
  const context: ParserContext = {
    tokens,
    options,
    result,
    isEpisodeKeywordsFound: false
  };

  searchForKeywords(context);
  searchForIsolatedNumbers(context);

  if (options.parseEpisodeNumber) {
    searchForEpisodeNumber(context);
  }

  searchForAnimeTitle(context);

  if (options.parseReleaseGroup && !hasResult(context, ElementCategory.ReleaseGroup)) {
    searchForReleaseGroup(context);
  }

  if (options.parseEpisodeTitle && hasResult(context, ElementCategory.EpisodeNumber)) {
    searchForEpisodeTitle(context);
  }

  validateElements(context);

  return { ok: hasResult(context, ElementCategory.AnimeTitle), result: context.result };
}

function searchForKeywords(context: ParserContext) {
  const tokens = context.tokens;
  const options = context.options;

  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    if (token.category !== TokenCategory.Unknown) continue;

    let word = trim(token.content, [' ', '-']);
    if (word === '') continue;

    // Don't bother if the word is a number that cannot be CRC
    if (word.length !== 8 && isNumericString(word)) continue;

    let category = ElementCategory.Unknown;
    const keyword = KeywordManager.normalize(word);
    const found = KeywordManager.find(keyword, ElementCategory.Unknown);
    if (found) {
      category = found.category;
      if (!options.parseReleaseGroup && category === ElementCategory.ReleaseGroup) {
        continue;
      }
      if (!isElementCategorySearchable(category) || !found.searchable) {
        continue;
      }
      if (isElementCategorySingular(category) && hasResult(context, category)) {
        continue;
      }

      switch (found.category) {
        case ElementCategory.AnimeSeasonPrefix:
          CheckAndSetAnimeSeasonKeyword(context, i);
          continue;
        case ElementCategory.EpisodePrefix:
          if (found.valid) {
            checkExtentKeyword(context, ElementCategory.EpisodeNumber, i);
          }
          continue;
        case ElementCategory.ReleaseVersion:
          word = word.slice(1);
          break;
        case ElementCategory.VolumePrefix:
          checkExtentKeyword(context, ElementCategory.VolumeNumber, i);
          continue;
      }
    } else {
      // CRC32 checksum
      if (!hasResult(context, ElementCategory.FileChecksum) && isCRC32(word)) {
        category = ElementCategory.FileChecksum;
      } else if (!hasResult(context, ElementCategory.VideoResolution) && isResolution(word)) {
        category = ElementCategory.VideoResolution;
      }
    }

    if (category !== ElementCategory.Unknown) {
      setResult(context, category, word);
      if (!found || found.identifiable) {
        token.category = TokenCategory.Identifier;
      }
    }
  }
}

function searchForIsolatedNumbers(context: ParserContext) {
  for (let i = 0; i < context.tokens.length; i++) {
    const token = context.tokens[i];
    if (
      token.category !== TokenCategory.Unknown ||
      !isNumericString(token.content) ||
      !isTokenIsolated(context, i)
    ) {
      continue;
    }

    // Anime Year
    const num = +token.content;
    if (AnimeYearMin <= num && num <= AnimeYearMax) {
      if (!hasResult(context, ElementCategory.AnimeYear)) {
        setResult(context, ElementCategory.AnimeYear, token.content);
        token.category = TokenCategory.Identifier;
        continue;
      }
    }

    // Video Resolution
    if (num !== 480 && num !== 720 && num !== 1080) continue;
    // If these numbers are isolated, it's more likely for them to be the
    // video resolution rather than the episode number. Some fansub groups use these without the "p" suffix.
    if (hasResult(context, ElementCategory.VideoResolution)) continue;
    setResult(context, ElementCategory.VideoResolution, token.content);
    token.category = TokenCategory.Identifier;
  }
}

function searchForEpisodeNumber(context: ParserContext) {}

function searchForAnimeTitle(context: ParserContext) {}

function searchForReleaseGroup(context: ParserContext) {}

function searchForEpisodeTitle(context: ParserContext) {}

function validateElements(context: ParserContext) {}
