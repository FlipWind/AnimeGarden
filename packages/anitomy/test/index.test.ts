import { describe, it, expect } from 'vitest';

import { parse } from '../src';

import { other } from './testcase';

describe('Other filenames', () => {
  it('should parse', () => {
    for (const filename of other) {
      expect(parse(filename)).toMatchSnapshot();
    }
  });
});
