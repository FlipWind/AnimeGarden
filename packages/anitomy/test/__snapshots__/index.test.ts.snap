// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > should work 1`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[Lilith-Raws] 熊熊勇闖異世界 PUNCH！ / Kuma Kuma Kuma Bear S02 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": undefined,
  "prefix": {
    "episode": undefined,
    "season": undefined,
    "volume": undefined,
  },
  "season": undefined,
  "subtitles": undefined,
  "title": undefined,
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
}
`;

exports[`parse > should work 2`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[ANi] 機動戰士鋼彈 水星的魔女 Season 2 - 13 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]",
  },
  "language": undefined,
  "prefix": {
    "episode": undefined,
    "season": undefined,
    "volume": undefined,
  },
  "season": undefined,
  "subtitles": undefined,
  "title": undefined,
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
}
`;

exports[`parse > should work 3`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[爱恋字幕社][4月新番][魔法使的新娘2][Mahoutsukai no Yome S2][01][1080P][MP4][GB][简中]",
  },
  "language": undefined,
  "prefix": {
    "episode": undefined,
    "season": undefined,
    "volume": undefined,
  },
  "season": undefined,
  "subtitles": undefined,
  "title": undefined,
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
}
`;

exports[`parse > should work 4`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[桜都字幕组] 总之就是非常可爱 S2 / Tonikaku Kawaii S2 [01][1080p][简体内嵌]",
  },
  "language": undefined,
  "prefix": {
    "episode": undefined,
    "season": undefined,
    "volume": undefined,
  },
  "season": undefined,
  "subtitles": undefined,
  "title": undefined,
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
}
`;

exports[`parse > should work 5`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "【喵萌奶茶屋】★04月新番★[百合是我的工作！/我的百合乃工作是也！/私の百合はお仕事です！/Watashi no Yuri wa Oshigoto desu!][01][720p][繁日雙語][招募翻譯校對]",
  },
  "language": undefined,
  "prefix": {
    "episode": undefined,
    "season": undefined,
    "volume": undefined,
  },
  "season": undefined,
  "subtitles": undefined,
  "title": undefined,
  "type": undefined,
  "video": {
    "resolution": "720p",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
}
`;
